test/test_multigrid_solver.jl

```diff
@@ -0,0 +1,53 @@
+include("dependencies_for_runtests.jl")
+
+using Oceananigans.Solvers: solve!
+using Statistics
+
+function run_poisson_equation_test(grid)
+    arch = architecture(grid)
+    # Solve ∇²ϕ = r
+    ϕ_truth = CenterField(grid)
+
+    # Initialize zero-mean "truth" solution with random numbers
+    set!(ϕ_truth, (x, y, z) -> rand())
+    parent(ϕ_truth) .-= mean(ϕ_truth)
+    fill_halo_regions!(ϕ_truth)
+
+    # Calculate Laplacian of "ϕ_truth"
+    ∇²ϕ = r = CenterField(grid)
+    compute_∇²!(∇²ϕ, ϕ_truth, arch, grid)
+
+    solver = MultigridSolver(compute_∇²!, arch, grid; template_field = ϕ_truth, reltol=eps(eltype(grid)))
+
+    # Solve Poisson equation
+    ϕ_solution = CenterField(grid)
+    solve!(ϕ_solution, solver, r)
+    ϕ_solution .-= mean(ϕ_solution)
```

Comment:
is this need @elise-palethorpe says?

Reply:
@elise-palethorpe seems that the matrix solver also requires that

https://github.com/CliMA/Oceananigans.jl/blob/3469cb714343019cedca9a12af9e21d4f9432e0d/test/test_matrix_poisson_solver.jl#L103

I can explain why in person?

Reply:
Wonderful, yes please! 

One other thing I noticed from the testing is the multigrid `solve!` method calls `fill_halo_regions!` 
https://github.com/CliMA/Oceananigans.jl/blob/3469cb714343019cedca9a12af9e21d4f9432e0d/src/Solvers/multigrid_solver.jl#L154
whereas all other solvers seem to do this separately - for example
https://github.com/CliMA/Oceananigans.jl/blob/3469cb714343019cedca9a12af9e21d4f9432e0d/test/test_preconditioned_conjugate_gradient_solver.jl#L41

Shall I change this?

Reply:
Yes, let's do what the other solvers do.

Reply:
Thanks, shall I also add the line
```x.-= mean(x)```
to `solve!` or is this only necessary in certain cases?

Reply:
No don't do that.

Do the other solvers do that?

Reply:
Removing the mean here is needed because of the laplacian operator. Essentially you can think that we are trying to solve the problem

```
d²φ/ dx² = b
```

This 2nd order problem has two free parameters. It requires two more constraints to be fully determined.

If the grid is periodic then we have 1 condition `φ(0) = φ(Lx)` so there is one free parameter left... It's the mean value of the solution in this case. If we have a 1D grid with `Bounded` topology, then there are two boundary conditions and this constrains the problem so you don't need to remove the mean.

```Julia
julia> grid = RectilinearGrid(arch, size=(4,), extent=(1.3,), topology = (Bounded, Flat, Flat))
4×1×1 RectilinearGrid{Float64, Bounded, Flat, Flat} on CPU with 3×0×0 halo
├── Bounded  x ∈ [-8.88178e-17, 1.3] regularly spaced with Δx=0.325
├── Flat y
└── Flat z

julia> ϕ_truth = CenterField(grid);

julia> set!(ϕ_truth, (x, y, z) -> rand());

julia> parent(ϕ_truth) .-= mean(ϕ_truth);

julia> ∇²ϕ = r = CenterField(grid);

julia> compute_∇²!(∇²ϕ, ϕ_truth, arch, grid)

julia> solver = MultigridSolver(compute_∇²!, arch, grid; template_field = ϕ_truth, reltol=eps(eltype(grid)));

julia> ϕ_solution = CenterField(grid)

julia> solve!(ϕ_solution, solver, r)

julia> fill_halo_regions!(ϕ_solution)

julia> ∇²ϕ_solution = CenterField(grid);

julia> compute_∇²!(∇²ϕ_solution, ϕ_solution, arch, grid)

julia> extrema_tolerance = 1e-12;

julia> std_tolerance = 1e-13;

julia> @test minimum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
Test Passed

julia> @test maximum(abs, interior(∇²ϕ_solution) .- interior(∇²ϕ)) < extrema_tolerance
Test Passed

julia> @test std(interior(∇²ϕ_solution) .- interior(∇²ϕ)) < std_tolerance
Test Passed

julia> @test minimum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
Test Passed

julia> @test maximum(abs, interior(ϕ_solution) .- interior(ϕ_truth)) < extrema_tolerance
Test Passed

julia> @test std(interior(ϕ_solution) .- interior(ϕ_truth)) < std_tolerance
Test Passed
```

Now for two or three dimensions I think even `Bounded` boundary conditions still will leave for a free parameter.

Reply:
Ahh thank you! That makes a lot of sense